### PR TITLE
Watson/subcerts

### PR DESCRIPTION
**Pull request by wbl**
This adds support for issuing delegated credentials as well as limiting a connection to only issue delegated credentials.

### CODE REVIEW

File: cmd/gokeyless-cli-tester/gokeyless-cli-tester.go [hunk 1]
```diff
@@ -50,6 +60,55 @@ func main() {
...
+		r, err := c.LookupServer(keyserver)
+		if err != nil {
+			log.Fatal("Could not look up keyserver", err)
+		}
+		conn, err := r.Dial(c)
+		if err != nil {
+			log.Fatal("Could not establish connection", err)
+		}
+		rpc := conn.RPC()
+		certfile, err := os.Open(delegateCert)
```
Comment (Bren2010) [inline cmd/gokeyless-cli-tester/gokeyless-cli-tester.go]: pemblock, err := ioutil.ReadFile(delegateCert) ?

File: cmd/gokeyless-cli-tester/gokeyless-cli-tester.go [hunk 2]
```diff
@@ -50,6 +60,55 @@ func main() {
...
+		}
+
+		ttl := 24 * time.Hour
+		cred, _, err := delegated.NewCredential(uint16(tls.ECDSAWithP256AndSHA256), delegated.VersionTLS13, ttl)
+		req, err := cred.Marshal()
+		if err != nil {
+			log.Fatal("failed to create query", err)
+		}
+		var resp []byte
+		err = rpc.Call("Delegator.Sign", req, &resp) //Richer queries and responses would be a good idea
```
Comment (Bren2010) [inline cmd/gokeyless-cli-tester/gokeyless-cli-tester.go]: Note 1: "// Richer" (space between)
Note 2: The note is right, you should be able to pass Credential / DelegatedCredential over RPC without marshalling/unmarshalling. Unless you chose not to do that for a reason?

File: cmd/gokeyless-cli-tester/gokeyless-cli-tester.go [hunk 3]
```diff
@@ -50,6 +60,55 @@ func main() {
...
+			log.Fatal("failed in RPC", err)
+		}
+		dc, err := delegated.UnmarshalDelegatedCredential(resp)
+		if err != nil {
+			log.Fatal("failed to parse response")
+		}
+		valid, err := dc.Validate(cert, time.Now())
+		if err != nil {
+			log.Fatal("Error in validation", err)
+		}
```
Comment (Bren2010) [inline cmd/gokeyless-cli-tester/gokeyless-cli-tester.go]: else if !valid

File: cmd/gokeyless-cli-tester/gokeyless-cli-tester.go [hunk 4]
```diff
@@ -50,6 +60,55 @@ func main() {
...
+		if !valid {
+			log.Fatal("Got invalid response")
+		}
+		log.Fatalf("Got valid response")
+		log.Fatalf("Got response %v", resp)
```
Comment (Bren2010) [inline cmd/gokeyless-cli-tester/gokeyless-cli-tester.go]: Only the first Fatalf is ever going to come through, and the script is going to exit with status 1. Should be 0 if successful.

fmt.Printf
fmt.Printf
return

File: cmd/gokeyless-cli-tester/gokeyless-cli-tester.go [hunk 5]
```diff
@@ -50,6 +60,55 @@ func main() {
 	if err != nil {
 		log.Fatal("failed to load cert directory:", err)
 	}
+	if delegate {
```
Comment (Bren2010) [inline cmd/gokeyless-cli-tester/gokeyless-cli-tester.go]: This is basically a standalone function, but you've wrapped it in an if statement. I'd move this code into its own function, and move the c.ScanDir call closer to `for _, priv := range privs` than the invocation of this.

**Comment by wbl** [comment]
I think I've addressed all the comments.

**Pull request merged**